File: src/main/scala/kim/dencka/java/easy/collection/SummaryRanges.java
package kim.dencka.java.easy.collection;

import java.util.ArrayList;
import java.util.List;

public class SummaryRanges {
    public static void main(String[] args) {
        int[] nums = {};
        System.out.println(summaryRanges(nums));
    }

    public static List<String> summaryRanges(int[] nums) {
        ArrayList<String> res = new ArrayList<>();
        ArrayList<Integer> range = new ArrayList<>();
        for (int i = 0; i < nums.length - 1; i++) {
            if (nums[i] + 1 == nums[i + 1]) {
                range.add(nums[i]);
            } else {
                if (range.size() == 0) {
                    res.add(nums[i] + "");
                } else {
                    res.add(range.get(0) + "->" + nums[i]);
                    range = new ArrayList<>();
                }
            }
        }
        if (range.size() == 0) {
            if (nums.length != 0) {
                res.add(nums[nums.length - 1] + "");
            }
        } else {
            res.add(range.get(0) + "->" + nums[nums.length - 1]);
        }
        return res;
    }
}

File: src/main/scala/kim/dencka/scala/easy/SummaryRanges.scala
package kim.dencka.scala.easy

object SummaryRanges extends App {

  def summaryRanges(nums: Array[Int]): List[String] = {
    @annotation.tailrec
    def run(arr: Array[Int], list: List[Int], acc: List[String]): List[String] = arr match {
      case Array() =>
        if (list.length == 1) acc :+ list.head.toString
        else acc :+ s"${list.head}->${list.last}"
      case _ if arr.head - 1 == list.last => run(arr.tail, list :+ arr.head, acc)
      case _ =>
        if (list.length == 1) run(arr.tail, List(arr.head), acc :+ list.head.toString)
        else run(arr.tail, List(arr.head), acc :+ s"${list.head}->${list.last}")
    }

    if (nums.isEmpty) List.empty
    else run(nums.tail, List(nums.head), List.empty[String])
  }
  println(summaryRanges(Array(0, 1, 2, 4, 6, 7, 8, 10)))
  println(summaryRanges(Array(0, 2, 4, 6, 8, 10, 12)))
  println(summaryRanges(Array()))

}
